internal/terraform/eval_conditions.go

```diff
@@ -126,7 +131,23 @@ func evalCheckRules(typ checkType, rules []*configs.CheckRule, ctx EvalContext,
 					EvalContext: hclCtx,
 				})
 			} else {
-				errorMessage = strings.TrimSpace(errorValue.AsString())
+				if marks.Has(errorValue, marks.Sensitive) {
+					diags = diags.Append(&hcl.Diagnostic{
+						Severity: severity,
+
+						Summary: "Error message refers to sensitive values",
+						Detail: `The error expression used to explain this condition refers to sensitive values. Terraform will not display the resulting message.
+
+You can correct this by removing references to sensitive values, or carefully using the nonsensitive() function if the expression will not reveal the sensitive data.`,
```

Comment:
This is probably just a "me" writing-style thing, so feel free to ignore if this doesn't feel right, but FWIW...

```suggestion
You can correct this by removing references to sensitive values, or by carefully using the nonsensitive() function if the expression will not reveal the sensitive data.`,
```